File: basics/topological_sort.cpp
#include <bits/stdc++.h>
using namespace std;
int N, M, a, b;
bool visited[200000];
vector<int> adj[200000];
vector<int> stackk;


void dfs(int curr){
  visited[curr] = true;

  for (auto node: adj[curr]){
    if (!visited[node]){
      dfs(node);
    }
  }
  stackk.push_back(curr);
}


int main(){
  cin >> N >> M;
  for (int i = 0; i < M; i++){
    cin >> a >> b;
    adj[a].push_back(b);
  }


  for (int i = 0; i < N; i++){
    if (!visited[i]){
      dfs(i);
    }
  }

  for (int i = N-1; i >= 0; i--){
    cout << stackk[i] << " ";
  }cout << "\n";
  
  return 0;
}   

File: basics/topological_sort.py
import sys
N, M = [int(x) for x in sys.stdin.readline().split()]
adj = [[] for x in range(N)]
for i in range(M):
  a, b = [int(x) for x in sys.stdin.readline().split()]
  adj[a].append(b)

visited = [False for x in range(N)]
stack = []

def dfs(curr):
  visited[curr] = True

  for node in adj[curr]:
    if not visited[node]:
      dfs(node)


  stack.append(curr)

for i in range(N):
  if (not visited[i]):
    dfs(i)

print(list(reversed(stack)))